packages/arb-avm-cpp/avm_values/include/avm_values/code.hpp

```diff
@@ -0,0 +1,194 @@
+                                           uint64_t pc) const {
+        return std::make_shared<CodeSegment>(
+            new_segment_id,
+            std::vector<CodePoint>{code.begin(), code.begin() + pc});
+    }
+
+   public:
+    CodeSegment(uint64_t segment_id_) : segment_id(segment_id_) {
+        code.push_back(getErrCodePoint());
+    }
+    CodeSegment(uint64_t segment_id_, std::vector<CodePoint> code_)
+        : segment_id(segment_id_), code(std::move(code_)) {}
+
+    uint64_t segmentID() const { return segment_id; }
+
+    const CodePoint& operator[](uint64_t pc) const { return code.at(pc); }
+
+    const CodePoint& at(uint64_t pc) const { return code.at(pc); }
+
+    friend std::ostream& operator<<(std::ostream& os, const CodeSegment& code);
+};
+
+struct CodeSegmentSnapshot {
+    std::shared_ptr<const CodeSegment> segment;
+    uint64_t op_count;
+};
+
+struct CodeSnapshot {
+    std::unordered_map<uint64_t, CodeSegmentSnapshot> segments;
+    uint64_t op_count;
+};
+
+class Code {
+    mutable std::mutex mutex;
+    std::unordered_map<uint64_t, std::shared_ptr<CodeSegment>> segments;
+    uint64_t next_segment_num;
+
+   public:
+    Code() : Code(0) {}
+    Code(uint64_t next_segment_num_) : next_segment_num(next_segment_num_) {}
+
+    std::shared_ptr<const CodeSegment> loadCodeSegment(
+        uint64_t segment_num) const {
+        const std::lock_guard<std::mutex> lock(mutex);
+        return segments.at(segment_num);
+    }
+
+    bool containsSegment(uint64_t segment_id) const {
+        const std::lock_guard<std::mutex> lock(mutex);
+        return segments.find(segment_id) != segments.end();
+    }
+
+    void restoreExistingSegment(std::shared_ptr<CodeSegment> segment) {
+        const std::lock_guard<std::mutex> lock(mutex);
+        uint64_t segment_id = segment->segmentID();
+        if (segments.find(segment->segmentID()) == segments.end()) {
+            segments[segment_id] = std::move(segment);
+        }
+    }
+
+    CodeSnapshot snapshot() const {
+        const std::lock_guard<std::mutex> lock(mutex);
+        std::unordered_map<uint64_t, CodeSegmentSnapshot> copied_segments;
+        for (const auto& key_val : segments) {
+            copied_segments[key_val.first] = {key_val.second,
+                                              key_val.second->size()};
+        }
+        return {std::move(copied_segments), next_segment_num};
+    }
+
+    const CodePoint& loadCodePoint(const CodePointRef& ref) const {
+        const std::lock_guard<std::mutex> lock(mutex);
+        return (*segments.at(ref.segment))[ref.pc];
+    }
+
+    CodePointStub addSegment() {
+        const std::lock_guard<std::mutex> lock(mutex);
+        uint64_t segment_num = next_segment_num++;
+        auto new_segment = std::make_shared<CodeSegment>(segment_num);
+        CodePointStub stub{{segment_num, 0}, hash((*new_segment)[0])};
+        segments[segment_num] = std::move(new_segment);
+        return stub;
+    }
+
```

Comment:
seems like these two functions can be made more dry, reusing logic to set `segments` values and incrementing next_segment_num. 

a function that constructs `CodePointStub stub{{segment_num, 0}, hash((*new_segment)[0])};` from a segment?

Reply:
That's a good call. I added a `codePointStub` function to `CodeSegment`